flaml/searcher/online_searcher.py

```diff
@@ -0,0 +1,371 @@
+        self._init_config = init_config
+        self._seed = config_oracle_random_seed
+        self._space = space
+        self._online_trial_args = online_trial_args
+        self._nonpoly_searcher_name = nonpoly_searcher_name
+
+        self._random_state = np.random.RandomState(self._seed)
+        self._searcher_for_nonpoly_hp = {}
+        self._space_of_nonpoly_hp = {}
+        # dicts to remember the mapping between searcher_trial_id and trial_id
+        self._searcher_trialid_to_trialid = {}  # key: searcher_trial_id, value: trial_id
+        self._trialid_to_searcher_trial_id = {}  # value: trial_id, key: searcher_trial_id 
+        self._challenger_list = []
+        # initialize the search in set_search_properties
+        self.set_search_properties(metric, mode, {self.CHAMPION_TRIAL_NAME: None}, init_call=True)
+        logger.debug('using random seed %s in config oracle', self._seed)
+
+    def set_search_properties(self, metric: Optional[str], mode: Optional[str],
+                              config: dict, init_call=False):
+        """Construct search space with given config, and setup the search
+        """
+        super().set_search_properties(metric, mode, config)
+        # *********Use ConfigOralce (i.e, self._generate_new_space to generate list of new challengers)
+        logger.info('champion trial %s', config)
+        champion_trial = config.get(self.CHAMPION_TRIAL_NAME, None)
+        if champion_trial is None:
+            champion_trial = self._create_trial_from_config(self._init_config)
+        # generate a new list of challenger trials
+        new_challenger_list = self._query_config_oracle(champion_trial.config,
+                                                        champion_trial.trial_id,
+                                                        self._trialid_to_searcher_trial_id[champion_trial.trial_id])
+        # add the newly generated challengers to existing challengers
+        # there can be duplicates and we check duplicates when calling next_trial()
+        self._challenger_list = self._challenger_list + new_challenger_list
+        # add the champion as part of the new_challenger_list when called initially
+        if init_call:
+            self._challenger_list.append(champion_trial)
+        logger.critical('Created challengers from champion %s', champion_trial.trial_id)
+        logger.critical('New challenger size %s, %s', len(self._challenger_list),
+                        [t.trial_id for t in self._challenger_list])
+
+    def next_trial(self):
+        """Return a trial from the _challenger_list
+        """
+        next_trial = None
+        if self._challenger_list:
+            next_trial = self._challenger_list.pop()
+        return next_trial
+
+    def _create_trial_from_config(self, config, searcher_trial_id=None):
+        if searcher_trial_id is None:
+            searcher_trial_id = Trial.generate_id()
+        trial = VWOnlineTrial(config, **self._online_trial_args)
```

Comment:
is the searcher specific to VW?
online_searcher and VWOnlineTrial do not follow consistent naming